src/Microsoft.IdentityModel.Abstractions/EventDetails.cs

```diff
@@ -0,0 +1,201 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.IdentityModel.Abstractions
+{
+    /// <summary>
+    /// Details of the telemetry event.
+    /// </summary>
+    /// <remarks>
+    /// This implementation is not meant to be thread-safe. This implementation would either need to be overridden or
+    /// usage should not be concurrently operated on.
+    /// </remarks>
+    public abstract class EventDetails
+    {
+        internal Dictionary<string, DataClassification> DataClassificationMapping = new Dictionary<string, DataClassification>();
+        internal Dictionary<string, string> PropertyValues = new Dictionary<string, string>();
+        internal Dictionary<string, bool> BoolPropertyValues = new Dictionary<string, bool>();
+        internal Dictionary<string, DateTime> DateTimePropertyValues = new Dictionary<string, DateTime>();
+        internal Dictionary<string, long> LongPropertyValues = new Dictionary<string, long>();
+        internal Dictionary<string, double> DoublePropertyValues = new Dictionary<string, double>();
+        internal Dictionary<string, Guid> GuidPropertyValues = new Dictionary<string, Guid>();
+        internal Dictionary<string, DictionaryTypeEnum> PropertiesKeys = new Dictionary<string, DictionaryTypeEnum>();
+
+        /// <summary>
+        /// Name of the telemetry event, should be unique between events.
+        /// </summary>
+        public abstract string Name { get; set; }
+
+        /// <summary>
+        /// Properties which describe the event.
+        /// </summary>
+        public virtual Dictionary<string, object> Properties
+        {
+            get
+            {
+                var currentBag = new Dictionary<string, object>();
```

Comment:
Depends on the implementation of ITelemetryClient really. For the version I created, it's just called once. The other option would be to store everything in a dictionary expose consumers to it directly. Main downside there is if we keep it object someone could put junk in that would break the ItelemetryClient so that seems worse dev ex. If someone was super concerned with perf (maybe this will even be us) they could take that usability tradeoff and override the implementation.

In practice most events will only have a half dozen or less dimensions so this should be relatively cheap. I certainly think you're right to be concerned when you see this though. Open to other ideas.

Reply:
Changed implementation to expose a readonly dictionary instead and avoided the looping.